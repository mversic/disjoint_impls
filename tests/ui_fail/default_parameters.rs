use disjoint_impls::disjoint_impls;

pub trait Dispatch {
    type Group;
}

pub trait Trait {}

pub enum GroupA {}
impl Dispatch for String {
    type Group = GroupA;
}
impl<T> Dispatch for Vec<T> {
    type Group = GroupA;
}

pub enum GroupB {}
impl Dispatch for i32 {
    type Group = GroupB;
}
impl Dispatch for u32 {
    type Group = GroupB;
}

impl Trait for u32 {}
impl<T> Trait for [T; 1] {}

disjoint_impls! {
    pub trait Kita<T: Trait = u32, const N: usize = 12> {
        const NAME: &'static str;
    }

    impl<T: Dispatch<Group = GroupA>> Kita for T {
        const NAME: &'static str = "Blanket A";
    }
    impl<T: Dispatch<Group = GroupA>> Kita<[T; 1]> for T {
        const NAME: &'static str = "Blanket B";
    }
}

fn main() {}
