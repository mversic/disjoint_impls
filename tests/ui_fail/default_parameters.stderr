error: Specify all trait arguments (including default)
  --> tests/ui_fail/default_parameters.rs:33:39
   |
33 |     impl<T: Dispatch<Group = GroupA>> Kita for T {
   |                                       ^^^^
